### DISCUSSION TITLE
Is the ngx_stream_ssl_preread_module of stream support the x25519kyber768draft00?

### DISCUSSION THREAD

**Discussion by QbsuranAlang** [nginx/nginx#469] [Q&A]
The chrome add a type of key share call `x25519kyber768draft00` which is very large. In my test, it has 1263 bytes.

It lets the ClientHello fragmented into multi segments.(At least 2 segments)

As I known, at this phase(SSL preread), nginx calls `read()` with `MSG_PEEK` once and only once. It must not call twice to get the next segment. Due to some feature of `MSG_PEEK` with kernel things.

It causes a lots of IPS, IDS, WAFm SSLi and so on cannot work as well.

Is nginx affected? I don't have environment to simulate.

**Comment by arut** [comment]
Why do you think `recv(MSG_PEEK)` can only be called once?

**Comment by QbsuranAlang** [comment]
> Why do you think `recv(MSG_PEEK)` can only be called once?

I meant `recv(MSG_PEEK)` cannot be call twice, or it will be get the same data again.

**Comment by QbsuranAlang** [comment]
```
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif /* _GNU_SOURCE */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/epoll.h>
#include <arpa/inet.h>

int stop = 0;

void handler(int sig) {
    stop = 1;
}//end handler

static void epoll_ctl_add(int epfd, int fd, uint32_t events) {
    struct epoll_event ev;

    ev.events = events;
    ev.data.fd = fd;
    if(epoll_ctl(epfd, EPOLL_CTL_ADD, fd, &ev) == -1) {
        perror("epoll_ctl()");
        exit(1);
    }//end if
}//end epoll_ctl_add

int main(int argc, char *argv[]) {

    /* open socket() */
    int srv_fd = socket(PF_INET, SOCK_STREAM, 0);
    if(srv_fd < 0) {
        perror("socket()");
        exit(1);
    }//end if

    /* bind() and listen() */
    struct sockaddr sa;
    struct sockaddr_in *in;
    in = (struct sockaddr_in *)&sa;
    memset(&sa, 0, sizeof(&sa));

    in->sin_family = PF_INET;
    in->sin_addr.s_addr = inet_addr("0.0.0.0");
    in->sin_port = htons(12345);

    int ret = bind(srv_fd, in, sizeof(sa));
    if(ret < 0) {
        perror("bind()");
        exit(1);
    }//end if

    ret = listen(srv_fd, 5);
    if(ret < 0) {
        perror("listen()");
        exit(1);
    }//end if

    /* epoll() jobs() */
    int ep = epoll_create1(EPOLL_CLOEXEC);
    if(ep < 0) {
        perror("epoll_create1()");
        exit(1);
    }//end if

    epoll_ctl_add(ep, srv_fd, EPOLLIN | EPOLLET);

    /* start epoll() */
    while(!stop) {
        struct epoll_event ee[1];
        ret = epoll_wait(ep, ee, 1, 0);
        if(ret < 0) {
            perror("epoll_wait()");
            exit(1);
        }//end if
        else if(ret == 0) {
            continue;
        }//end if
        else {
            for(int i = 0 ; i < ret ; i++) {

                /* server accept client */
                if(ee->data.fd == srv_fd) {
                    struct sockaddr sa;
                    struct sockaddr_in *in;
                    socklen_t len;

                    in = (struct sockaddr_in *)&sa;
                    memset(&sa, 0, sizeof(sa));

                    int fd = accept4(srv_fd, &sa, &len, SOCK_NONBLOCK);
                    if(fd < 0) {
                        perror("accept4()");
                        exit(1);
                    }//end if

                    epoll_ctl_add(ep, fd, EPOLLIN | EPOLLET);
                }//end if
                else {
                    int fd = ee->data.fd;


                    char msg[1500];
                    memset(msg, 0, sizeof(msg));
                    ssize_t nread = recv(fd, msg, sizeof(msg), MSG_PEEK);
                    printf("nread = %zd, msg =\n", nread);
                    for(int j = 0 ; j < 15 || j < strlen(msg) ; j++) {
                        printf("%#06x \n", msg[j]);
                    }//end for
                    printf("\n");

                    close(fd);
                    close(srv_fd);
                    exit(0); /* just try one client */
                }//end else
            }//end if
        }//end else
    }//end while

    close(srv_fd);
    return 0;
}//end main
```

If I didn't misunderstand the `ngx_stream_ssl_preead_module` logic is similar as the sample code.

```
$ ./a
nread = 1500, msg =
0x0016 
0x0003 
0x0001 
0x0007 
000000 
0x0001 
000000 
0x0006 
0xfffffffc 
0x0003 
0x0003 
0xffffffd0 
0xffffffdc 
0xffffff93 
0xffffffd2 

nread = 1500, msg =
0x0016 
0x0003 
0x0001 
0x0007 
000000 
0x0001 
000000 
0x0006 
0xfffffffc 
0x0003 
0x0003 
0xffffffd0 
0xffffffdc 
0xffffff93 
0xffffffd2 
```

Run the sample code, and let chrome run the target URL, the result will show the `recv(MSG_PEEK)` get the same data.
It didn't get the second segment data.

**Comment by Maryna-f5** [comment]
Moving the issue to Discussions

**Comment by Hi-Angel** [comment]
Hi @QbsuranAlang, occasionally stumbled upon your question

> Due to some feature of `MSG_PEEK` with kernel things.

I mean, that's the whole point of the flag, it is literally documented in `man 3 recv` as:

```
Peeks at an incoming message. The data is treated as unread and the next recv()
or similar function shall still return this data.
```

> It must not call twice to get the next segment.

Well… it actually depends.

What flag `MSG_PEEK` does is it allows to immutably peek at the kernel buffer, and the kernel buffer may receive more data, in which case subsequent calls with `MSG_PEEK` would allow you to see both older and newer data.

IOW, suppose you got sent 2 packets of 512 bytes size. Let's say you call `recv(fd, buf, 1024, MSG_PEEK)` twice: once after first packets comes in, and another one after second one comes in. In that case 1st call would return 512, and the second call 1024 *(because the older 512 bytes are still in the buffer, and now the buffer holds both packets)*.

---------

That said, I don't know nginx code, so neither I know if nginx processes that correctly *(i.e. whether it waits for more data if it sees the buffer doesn't have enough)*. If you can reproduce a problem with this *(e.g. by crafting a TLS hello and breaking it down to two parts with `sleep(1)` in between)*, you should report a bug.

FWIW, `multisplit` is a popular anti-DPI technique, which some nginx versions seem to handle correctly. So e.g. I can open nginx.org just fine *(which is currently running 1.29.8)*, but I can't open archlinux.org with multisplit *(which doesn't report its nginx version, so 🤷‍♂️)*
